File: 124-binary-tree-maximum-path-sum/binary-tree-maximum-path-sum.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int maxPathSum(TreeNode* root) {
        int maxi = INT_MIN;
        maxPath(root, maxi);
        return maxi;   
    }
    int maxPath(TreeNode* node, int &maxi){
        if(node == NULL)return 0;
        int left_sum = max(0, maxPath(node->left, maxi));
        int right_sum = max(0, maxPath(node->right, maxi));
        maxi = max(maxi, left_sum + right_sum + node->val);
        return max(left_sum, right_sum) + node->val;
    }
};

File: 124-binary-tree-maximum-path-sum/binary-tree-maximum-path-sum.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def maxPathSum(self, root: Optional[TreeNode]) -> int:
        self.maxi = float('-inf')

        def maxSum(node, m):
            if not node:
                return 0
            left = max(0, maxSum(node.left, self.maxi))
            right = max(0, maxSum(node.right, self.maxi))
            self.maxi = max(self.maxi, left + right + node.val)
            return max(left, right) + node.val
        maxSum(root, self.maxi)
        return self.maxi